### PR TITLE
Create a blast radius reduction guide

### LINKED ISSUE #9055: "Reducing the blast radius" docs page
## Details
Per a meeting with @klizhentas, this will be a page within a new `/docs/setup` section related to security. It will focus on how to limit the potential damage any Teleport security breaches. Each topic will explain a best practice for setting up Teleport to contain certain kinds of attack, and will describe the attack vectors it will protect your Teleport deployment from. 

Topics include:
- Making MFA mandatory
- Turn on per-session MFA for everything
- Dual-authorization administrative access set up for root accounts. Uses MFA and requires approval from two people to assume that role.
- Set up deny rules in Teleport RBAC that will prohibit anyone who is not from a team from ever becoming, or even requesting, an admin role.

### PR DESCRIPTION
**Pull request by ptgott**
This is the first guide within the "Security" subsection within
the "Setup" section of the docs site.

Closes #9055

### CODE REVIEW

File: docs/pages/setup/security.mdx [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+title: Security
+description: How to prevent attacks on your Teleport cluster
```
Comment (klizhentas) [inline docs/pages/setup/security.mdx:3]: I'd use something more generic: Hardening security of your Teleport deployment

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+Teleport encourages users to practice defense in depth so that every component of their infrastructure is protected against an attack, even if the attack is partially successful. You can configure Teleport to add layers of protection to your cluster when users authenticate or request elevated privileges. In this guide, we will show you how to:
+
+- [Make MFA mandatory for `tsh` logins](#make-mfa-mandatory-for-tsh-login)
+- [Present an MFA challenge for every attempt to access a resource](#present-an-mfa-challenge-for-every-attempt-to-access-a-resource)
+- [Require multiple reviews for role requests](#require-multiple-reviews-for-role-requests)
+- [Automatically prevent some roles from requesting others](#automatically-prevent-some-roles-from-requesting-others)
+
+
+## Make MFA mandatory for `tsh login`
+If a user sets up an account to authenticate to their Teleport cluster with only a password, an adversary can [gain access](/blog/no-passwords) to the password using brute-force attacks, person-in-the-middle attacks, or phishing. But even if a user's password is compromised, you can stop an attacker from authenticating with it when they run `tsh login`.
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:15]: nit: i'd remove this link to keep the readers flow. Take a look at some tips here:

https://www.divio.com/

and

https://goteleport.com/docs/docs/

We can add useful links in "Next steps" last section

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+- [Present an MFA challenge for every attempt to access a resource](#present-an-mfa-challenge-for-every-attempt-to-access-a-resource)
+- [Require multiple reviews for role requests](#require-multiple-reviews-for-role-requests)
+- [Automatically prevent some roles from requesting others](#automatically-prevent-some-roles-from-requesting-others)
+
+
+## Make MFA mandatory for `tsh login`
+If a user sets up an account to authenticate to their Teleport cluster with only a password, an adversary can [gain access](/blog/no-passwords) to the password using brute-force attacks, person-in-the-middle attacks, or phishing. But even if a user's password is compromised, you can stop an attacker from authenticating with it when they run `tsh login`. 
+
+For clusters that use the [local authentication connector](/docs/setup/reference/authentication/#local), Teleport lets you make it mandatory for a user to enroll an MFA device when they create an account, and to authenticate using that device when they begin a new Teleport session. To do so, add the following to your Teleport [configuration file](/docs/setup/reference/config):
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:17]: Local references have the following format: `[reference text](./path/to/file.mdx#anchor)`

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+
+## Present an MFA challenge for every attempt to access a resource
+After a user logs into a Teleport cluster, they can request access to a particular resource, e.g., a node, database, application, or Kubernetes cluster. In this case, the Teleport [Auth Service](/docs/architecture/authentication/) issues a single-use certificate for accessing that resource. You can prevent attackers from doing damage with a compromised certificate by enabling [per-session MFA](/docs/access-controls/guides/per-session-mfa/). 
+
+With this setting, whenever a user requests a one-time certificate to access a resource, the Teleport Auth Service will issue an MFA challenge, even if the user has already begun a Teleport session via `tsh login`. To enable per-session MFA for all users, make the following change to your Teleport configuration file:
+
+```yaml
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:48]: nit: this is only possible for users in the self-hosted mode. Let's use details with scope and tabs with scope to have references for cloud and self hosted. Cloud users can use RBAC options to achieve this effect.

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+
+## Require multiple reviews for role requests
+Even if an attacker gains access to a user's credentials and successfully signs into your Teleport cluster, you can still prevent the user from escalating their privileges. If you enable [dual authorization](/docs/access-controls/guides/dual-authz/), users who request to assume a particular role must obtain permission to do so from two or more reviewers. This way, if a malicous user manages to impersonate a legitimate one, reviewers can contact the real user before granting the new role.
+
+Dual authorization uses Teleport's [access plugins](/docs/enterprise/workflow/)—e.g., Slack, JIRA, and PagerDuty—to notify reviewers that a user has requested a role. Plugins are applications that communicate with your Teleport cluster via [gRPC](https://grpc.io/). Teleport's Auth Service broadcasts role request notifications to your access plugins, which handle the messages and notify users. For access plugins that require a SAML or OIDC [connector](/docs/enterprise/sso/#configuring-sso), you must enable the Cloud or Enterprise versions of Teleport.
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:60]: I think this is goes too deep in the implementations. I would simply mention that users can configure external systems to simplify collaboration and review here. I would add a section in next steps for folks to read more about plugins

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+- [The reviewee](#the-reviewee)
+
+### The reviewer
+You can enable some users to review other users' role escalation requests by applying a role template similar to the following:
+
+```yaml
+kind: role
+version: v4
+metadata:
+  name: <role name>
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:74]: I would remove angle brackets to avoid confusing our linter, I'd use simple name "reviewer"

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+spec:
+  deny:
+    request:
+      roles: ['admin']
+
+```
+
+Next, `myuser` attempts to request the `admin` role.
+
+```
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:124]: let's use code block here (described in docs/docs)

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+tsh request create --roles=admin
+```
+
+However, the Auth Service denies the request.
+
+```
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:130]: same

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+root@myhost:/# tsh request create --roles=admin
+Creating request...
+ERROR: user "myuser" can not request role "admin"
+```
```
Comment (klizhentas) [inline docs/pages/setup/security/reduce-blast-radius.mdx:134]: I would also add last step, using deny rules to prevent accidental privilege extension. You can setup an SSO role mapping that always maps all users who are not in special group "admins" to a role that will deny certain admin actions (like editing role, connector resources and auth preferences). So if someone makes a mistake, this deny rule will override allow rules, limiting accidental scope expansion.

**Review by klizhentas — COMMENTED** [review]
I think this article is well written, because it starts with potential attack vector and shows how to protect from it. I've added a bunch of mostly stylistic comments and suggest we add one more scenario. But the article is almost ready.

**Comment by klizhentas** [comment]
@r0mant @fspmarshall @russjones can you folks please review when you have a chance

**Comment by ptgott** [comment]
@klizhentas Thanks for your review! I'll push changes once I receive another review so I can address overlapping comments at the same time.

File: docs/pages/setup/security.mdx (hunk 1 above)
Comment (ptgott) [reply docs/pages/setup/security.mdx:3]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 2 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:15]: I'll move most inline links to a "next steps" section.

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 3 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:17]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 4 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:48]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 5 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:60]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 6 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:74]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 7 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:124]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 8 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:130]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 9 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:134]: Sounds good. Does this refer to the `teams_to_logins` field in the `github` resource? (https://goteleport.com/docs/setup/admin/github-sso/#step-12-create-github-connector)

**Comment by ptgott** [comment]
@klizhentas Aside from [one question](https://github.com/gravitational/teleport/pull/9189#discussion_r764213894), I have responded to your review.

@r0mant @fspmarshall @russjones would any of you have time to review this? Thanks!

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+```yaml
+kind: role
+version: v4
+metadata:
+  name: user
+spec:
+  deny:
+    request:
+      roles: ['admin']
+
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:160]:
```suggestion
```

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+<Tabs>
+<TabItem label="Self-hosted" scope={["oss","enterprise"]}>
+Add the following to your Teleport configuration file:
+
+```yaml
+auth_service:
+    authentication:
+        type: local
+        second_factor: otp|u2f|webauthn|on
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:28]: Nit: Use 2-space indentation for consistency.
```suggestion
auth_service:
  authentication:
    type: local
    second_factor: otp|u2f|webauthn|on
```

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  type: local
+  second_factor: otp|u2f|webauthn|on
+```
+
+Remember to create your dynamic resource using `tctl create -f <path to your YAML file>`.
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:44]:
```suggestion
Create your dynamic resource using `tctl create -f <path to your YAML file>`.
```

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+To make MFA mandatory for all users, `second_factor` must be set to one of the following values:
+
+- `otp`
+- `u2f`
+- `webauthn`
+- `on`
+
+Choose `on` if you would like to require MFA for all users while letting them choose an OTP, U2F, or WebAuthn device. The other options restrict users to a single type of MFA device, which is useful for enforcing a particular standard of security. Once you start the Teleport Proxy Service with the `second_factor` configuration option set to one of these values, Teleport will mandate MFA by:
+
+- Adjusting the Teleport signup page so a user must enroll an MFA device of the kind you have selected. If the value of `second_factor` is `on`, users will have the option to select from multiple device types.
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:57]: Let's explain the process for enabling an MFA device for existing users when the cluster is updated from "MFA off" to "MFA on". I think we also have an `optional` setting which admins can use during the grace period.

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+To enable per-session MFA for all users, do the following depending on your Teleport environment:
+
+<Tabs>
+<TabItem label="Self-hosted" scope={["oss","enterprise"]}>
+Make the following changes to your Teleport configuration file:
+
+```yaml
+auth_service:
+    authentication:
+        require_session_mfa: yes
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:75]:
```suggestion
auth_service:
  authentication:
    require_session_mfa: yes
```

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  require_session_mfa: yes
+```
+
+Remember to create your dynamic resource using `tctl create -f <path to your YAML file>`.
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:91]:
```suggestion
Create your dynamic resource using `tctl create -f <path to your YAML file>`.
```

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+</TabItem>
+</Tabs> 
+
+The `require_session_mfa` setting only applies to users who have already established a Teleport session. To require MFA for `tsh login`, see the [previous section](#make-mfa-mandatory-for-tsh-login).
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:95]: > The `require_session_mfa` setting only applies to users who have already established a Teleport session.

I'm not sure what this is referring to. Maybe rephrase somehow?

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+
+## Require multiple reviews for role requests
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:98]: Nit: Just call it "Dual authorization"?

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+      thresholds:
+        - approve: 2
+          deny: 1
+```
+
+The `spec.allow.request.roles` field lists the names of other roles that a user with the `reviewee` role can request. When a reviewee requests access to one of these roles, Teleport notifies reviewers via your access plugins. The `spec.allow.requests.roles.thresholds` field indicates how many reviews are required to approve or deny the request.
+
+
+## Automatically prevent some roles from requesting others
+In theory, a malicious Teleport user could request a more privileged role and trick a reviewer into granting access. You can prevent such a scenario by defining roles that prohibit users from even requesting access to particular roles.
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:145]: <strike>In theory</strike>

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+
+The `spec.deny` field has the same possible properties as the `spec.allow` field we described [earlier](#require-multiple-reviews-for-role-requests) except, rather than enabling actions, this field disables them. For example, the `spec.deny.requests.roles` field is a list of roles that a user is prohibited from requesting.
```
Comment (r0mant) [inline docs/pages/setup/security/reduce-blast-radius.mdx:147]: Might be helpful to mention here that "deny" rules take precedence if a user has multiple roles assigned.

**Review by r0mant — COMMENTED** [review]
Nice article! I left some comments. I was also wondering - would "session locking" be in scope of this article or is it intended to be more "configuration" oriented?

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 9 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 20]
```diff
@@ -0,0 +1,187 @@
+---
+title: Reducing the Blast Radius of Attacks
+description: You can easily configure Teleport to minimize the scope of security breaches
```
[the remark this answers is not in the document]
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:3]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 10 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:160]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 11 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:28]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 12 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:44]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 14 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:75]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 15 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:91]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 18 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:145]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 17 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:98]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 19 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:147]: 👍

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 16 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:95]: This sentence restates information that the page introduces earlier, so I think it should be fine to remove it.

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 13 above)
Comment (ptgott) [reply docs/pages/setup/security/reduce-blast-radius.mdx:57]: 👍

**Comment by ptgott** [comment]
@r0mant Thanks for the review! I have pushed changes. @klizhentas I have also added the sections we talked about. Would it make sense to add a section on session logging as well?

File: docs/pages/setup/security/reduce-blast-radius.mdx [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+      thresholds:
+        - approve: 2
+          deny: 1
+```
+
+The `claims_to_roles` field within an `allow` or `deny` rule maps a user's `traits` to `roles` that they are either permitted or forbidden to request. In this case, we use the `{{regexp.not_match(\"admin\")}}` template function to prevent any user  from requesting the `db-writer` role unless they have a `groups` trait with a value like `administrator` or `admins`. Users who _do_ have such a trait can request the role with two approvals.
+
+
+## Set up your RBAC without admin roles
+You can design your Teleport RBAC so that there is no all-powerful administrator in the system, or even a `reviewer` role with elevated privileges. This way, you can reduce the blast radius if an attacker successfully impersonates a Teleport user and requests a more privileged role.
```
Comment (fspmarshall) [inline docs/pages/setup/security/reduce-blast-radius.mdx:226]: Bit of a silly nitpick, but I feel like the example used in this section undermines the point being made here a bit.  The example below shows a user role that can only be partially escalated to editor, which seems to imply the existence of an admin-like actor with full privileges (the thing we're saying you can choose to avoid).

I think having separate examples for the "democratic" and "partial escalation" scenarios would make the points clearer and more memorable.

**Review by fspmarshall — APPROVED** [review]

File: docs/pages/setup/security/reduce-blast-radius.mdx (hunk 21 above)
Comment (klizhentas) [reply docs/pages/setup/security/reduce-blast-radius.mdx:226]: @fspmarshall @ptgott with access requests you can have 0 admins in the system at any given time. However when someone gets the privilege, they would become admin temporarily. We can limit what they can admin exactly, but that is only partial mitigation

**Comment by klizhentas** [comment]
@ptgott let's merge this

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
@ptgott let's also backport to branch/v8
